src/indentblock.js

```diff
@@ -0,0 +1,289 @@
+
+		editor.config.define( 'indentBlock', {
+			offset: 1,
+			unit: 'em'
+		} );
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	static get pluginName() {
+		return 'IndentBlock';
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	init() {
+		const editor = this.editor;
+		const schema = editor.model.schema;
+		const conversion = editor.conversion;
+		const configuration = editor.config.get( 'indentBlock' );
+
+		// TODO: supported in CKE4:
+		//  - p,
+		//  - h1, h2, h3, h4, h5, h6,
+		//  - ul, ol,
+		//  - div,
+		//  - dl,
+		//  - pre,
+		//  - table
+		const knownElements = [ 'paragraph', 'heading1', 'heading2', 'heading3', 'heading4', 'heading5', 'heading6' ];
+
+		knownElements.forEach( elementName => {
+			if ( schema.isRegistered( elementName ) ) {
+				schema.extend( elementName, { allowAttributes: 'blockIndent' } );
+			}
+		} );
+
+		const useOffsetConfig = !configuration.classes || !configuration.classes.length;
+
+		const indentConfig = Object.assign( { direction: 'forward' }, configuration );
+		const outdentConfig = Object.assign( { direction: 'backward' }, configuration );
+
+		if ( useOffsetConfig ) {
+			this._setupConversionUsingOffset( conversion );
+			editor.commands.add( 'indentBlock', new IndentBlockCommand( editor, new IndentUsingOffset( indentConfig ) ) );
+			editor.commands.add( 'outdentBlock', new IndentBlockCommand( editor, new IndentUsingOffset( outdentConfig ) ) );
+		} else {
+			this._setupConversionUsingClasses( configuration.classes );
+			editor.commands.add( 'indentBlock', new IndentBlockCommand( editor, new IndentUsingClasses( indentConfig ) ) );
+			editor.commands.add( 'outdentBlock', new IndentBlockCommand( editor, new IndentUsingClasses( outdentConfig ) ) );
+		}
+
+		const getCommandExecuter = commandName => {
+			return ( data, cancel ) => {
+				const command = this.editor.commands.get( commandName );
+
+				if ( command.isEnabled ) {
+					this.editor.execute( commandName );
+					cancel();
+				}
+			};
+		};
+
+		editor.keystrokes.set( 'Tab', getCommandExecuter( 'indentBlock' ) );
+		editor.keystrokes.set( 'Shift+Tab', getCommandExecuter( 'outdentBlock' ) );
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	afterInit() {
+		const editor = this.editor;
+		const indentCommand = editor.commands.get( 'indent' );
+		const outdentCommand = editor.commands.get( 'outdent' );
+
+		indentCommand.registerChildCommand( editor.commands.get( 'indentBlock' ) );
+		outdentCommand.registerChildCommand( editor.commands.get( 'outdentBlock' ) );
+	}
+
+	/**
+	 * Setups conversion for using offset indents.
+	 *
+	 * @private
+	 */
+	_setupConversionUsingOffset() {
+		const conversion = this.editor.conversion;
+
+		conversion.for( 'upcast' ).attributeToAttribute( {
+			view: {
+				styles: {
+					'margin-left': /[\s\S]+/
+				}
+			},
+			model: {
+				key: 'blockIndent',
+				value: viewElement => viewElement.getStyle( 'margin-left' )
+			}
+		} );
+
+		// The margin shorthand should also work.
+		conversion.for( 'upcast' ).attributeToAttribute( {
```

Comment:
We talked that style normalization should work on element/consumable level (so, be completely automatic), but if this works already, we can keep it.